### PR TITLE
Support edge features in StellarGraph objects

### PR DESCRIPTION
**Pull request by huonw**
This adds support for storing edge features in StellarGraph objects. This works by taking any non-special columns of the edges DataFrame(s) as feature columns, similar to nodes (except nodes have no special columns). In particular, features are columns other than `source_column`, `target_column`, `edge_weight_column` and (if specified) `edge_type_column`.

This opens up the ability for us to have algorithms that use edge information, such as #1328 and #1329.

This diff here is quite messy, but there's a few core changes:

- moving the "features" support into `ElementData` instead of `NodeData` by changing `type_starts` that was previously `[(type_name, first_type_iloc), ...]` into `type_info` that is `[(type_name, type_features), ...]` the length (of the first axis) of the `type_features` array thus defines which nodes are of that type
- making sure that the edges conversion code can create the appropriate NumPy arrays
- duplicating `unique_node_type`, `node_feature_sizes`, `node_features` to the `edge` equivalent, by creating private versions that switch between them
- updating the docs, including the Pandas demo

See: #1327

### CODE REVIEW

File: stellargraph/core/convert.py [hunk 1]
```diff
@@ -119,37 +111,27 @@ def select_column(old_name):
...
-            )
+        # to_numpy returns an unspecified order but it's Fortran in practice. Row-level bulk
+        # operations are more common (e.g. slicing out a couple of row, when sampling a few
+        # nodes) than column-level ones so having rows be contiguous (C order) is much more
+        # efficient.
+        features = np.ascontiguousarray(other.to_numpy(dtype=self.dtype))
 
         return ids, columns, features
 
     def _ids_columns_and_starts_from_singles(self, singles):
```
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/core/convert.py:122]: Good catch. 👍

File: stellargraph/core/graph.py [hunk 2]
```diff
@@ -1040,6 +1084,71 @@ def node_ilocs_to_ids(self, node_ilocs):
...
+            # empty lists are cast to a default array type of float64 -
+            # must manually specify integer type if empty, in which case we can pretend we received ilocs
+            ilocs = np.asarray(ids, dtype=np.uint8)
+            use_ilocs = True
+        elif use_ilocs:
+            ids = np.asarray(ids)
+            ilocs = np.asarray(ids)
+        else:
+            ids = np.asarray(ids)
+            ilocs = element_data.ids.to_iloc(np.asarray(ids))
```
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/core/graph.py:1108]: All of these were inherited from `node_features` and they do indeed seem rather unnecessary.

**Comment by huonw** [comment]
> I did briefly wonder if it would make more sense now for TemporalRandomWalk to use a single edge feature instead of edge weights once this lands, but I'm not sure if that's actually better.

Yeah, that's an interesting thought. I guess one could have it be a function of edge weights and edge features, which defaults to just returning the edge weights for backwards compatibility, but would open it to more complicated things.

File: stellargraph/core/element_data.py [hunk 3]
```diff
@@ -121,40 +121,60 @@ class ElementData:
...
-                    f"type_starts (for {type_name!r}): expected valid type range, found start ({start}) after stop ({stop})"
+                    f"type_info (for {type_name!r}): expected numpy array, found {type(data).__name__}"
                 )
+
+            if len(data.shape) != 2:
+                raise ValueError(
+                    f"type_info (for {type_name!r}): expected 2 dimensions, found {len(data.shape)}"
+                )
+
+            rows, _columns = data.shape
```
Comment (kjun9) [inline stellargraph/core/element_data.py:152]: I think `_columns` is unused here
```suggestion
            rows = data.shape[0]
```

Comment (huonw) [reply stellargraph/core/element_data.py:152]: It was on purpose in the original code (that's what the leading _ conveys, in my mind), trying to express the rank-2 nature of `data`... however, that was all changed by the multidimensional feature change in #1561 that's now merged, and I guess it'll be resolved by the merge with develop.

**Pull request merged**